### PR TITLE
RATIS-989. Avoid change state from CLOSING to EXCEPTION in LogAppender

### PR DESCRIPTION
**Pull request by runzhiwang**
## What changes were proposed in this pull request?

1. allow EXCEPTION to CLOSING by 
image

2. check LifeCycle.State.isValid(from, to) before state transition.

3. add synchronized to lifeCycle

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-989

## How was this patch tested?

Existed tests.

### CODE REVIEW

**Comment by runzhiwang** [comment]
@lokeshj1703 Could you help review this patch ? Thank you very much.

**Comment by runzhiwang** [comment]
CI can not be triggered, We should wait for CI.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LogAppender.java [hunk 1]
```diff
@@ -117,6 +121,16 @@ void stop() {
     public String toString() {
       return name;
     }
+
+    private boolean transLifeCycle(LifeCycle.State to) {
```
Comment (lokeshj1703) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LogAppender.java:125]: NIT: Lets rename it to transitionLifeCycle.

Comment (runzhiwang) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/LogAppender.java]: @lokeshj1703 Thanks for review. I have updated it.

**Comment by lokeshj1703** [comment]
@runzhiwang Thanks for the contribution! I have committed the PR to master branch.

**Pull request closed without merging**